File: apps/android/app/src/main/java/com/exponential/app/ui/components/picker/IconPicker.kt
package com.exponential.app.ui.components.picker

import com.exponential.app.domain.DomainContract

/**
 * EXP-1029 contract — `IconPicker(set)` under the shared picker API.
 *
 * `ui/components/IconPicker.kt` (EXP-575: the square trigger over
 * `IconSwatchGrid`; EXP-924: the SET is a parameter — the board set or the
 * device set) already IS the typed picker of this contract; EXP-1021
 * re-homes it onto [Picker] (the grid as the sheet's body, the trigger
 * unchanged). This file names the two sets as picker items so the contract
 * test can name the picker beside the other nine.
 */
enum class IconPickerSet {
    /** The 96 board / action glyphs (`icons.json` `pickable`). */
    Board,
    /** The six device glyphs (`icons.json` `devicePickable`). */
    Device,
}

/** The set's glyph names as picker items (the glyph is the icon itself). */
fun iconPickerItems(set: IconPickerSet): List<PickerItem<String>> {
    val names = when (set) {
        IconPickerSet.Board -> DomainContract.boardIconValues
        IconPickerSet.Device -> DomainContract.deviceIconValues
    }
    return names.map { PickerItem(value = it, label = it) }
}

File: apps/desktop/crates/ui/src/picker/icon_picker.rs
//! EXP-1029 contract — `IconPicker(set)` under the shared picker API: the
//! icon SET is the parameter (the 96 board/action glyphs or the six device
//! glyphs, `icons.json` `pickable` / `devicePickable`). `board_form::
//! icon_picker` (EXP-575, the square trigger over the swatch grid) moves onto
//! [`Picker`] in EXP-1021 with the grid as the surface's body.

use gpui::AnyElement;

use super::{OnPickerChange, Picker, PickerItem};

/// Which registry set the grid offers.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum IconSet {
    Board,
    Device,
}

pub(crate) fn icon_items(set: IconSet) -> Vec<PickerItem<String>> {
    let names: &[&str] = match set {
        IconSet::Board => domain::contract::BOARD_ICON_VALUES,
        IconSet::Device => domain::contract::DEVICE_ICON_VALUES,
    };
    names
        .iter()
        .map(|name| PickerItem::new((*name).to_string(), *name))
        .collect()
}

pub(crate) fn icon_picker(
    set: IconSet,
    value: Option<String>,
    trigger: AnyElement,
    on_change: OnPickerChange<String>,
) -> Picker<String> {
    Picker::single(icon_items(set), value, trigger, on_change)
}
